browser/src/e2e/github.test.ts

```diff
@@ -1,180 +1,115 @@
-        head: { text: 'typ', selector: 'span.pl-s1' },
+        // https://github.com/gorilla/mux/pull/117/files#diff-9ef8a22c4ce5141c30a501c542fb1adeL244
+        base: {
+            token: 'varsN',
+            lineId: 'diff-9ef8a22c4ce5141c30a501c542fb1adeL244',
+            goToDefinitionURL:
+                'https://github.com/gorilla/mux/blob/f15e0c49460fd49eebe2bcc8486b05d1bef68d3a/regexp.go#L139:2',
+        },
+        // https://github.com/gorilla/mux/pull/117/files#diff-9ef8a22c4ce5141c30a501c542fb1adeR247
+        head: {
+            token: 'host',
+            lineId: 'diff-9ef8a22c4ce5141c30a501c542fb1adeR247',
+            goToDefinitionURL:
+                'https://github.com/gorilla/mux/blob/e73f183699f8ab7d54609771e1fa0ab7ffddc21b/regexp.go#L233:2',
+        },
     }
 
-    for (const diffType of ['unified', 'split']) {
-        for (const side of ['base', 'head'] as const) {
-            it(`provides tooltips for diff files (${diffType}, ${side})`, async () => {
-                await page.goto(`https://github.com/gorilla/mux/pull/328/files?diff=${diffType}`)
-
-                const token = tokens[side]
-                const element = await getTokenWithSelector(page, token.text, token.selector)
-
-                // Scrolls the element into view so that code view is in view.
-                await element.hover()
-                await page.waitForSelector('[data-path="regexp.go"] .code-view-toolbar .open-on-sourcegraph')
-                await clickElement(page, element)
-                await page.waitForSelector('.e2e-tooltip-go-to-definition')
+    describe('Pull request pages', () => {
+        for (const diffType of ['unified', 'split']) {
+            describe(`${startCase(diffType)} view`, () => {
+                for (const side of ['base', 'head'] as const) {
+                    const { token, lineId } = tokens[side]
+                    it(`provides hover tooltips on token "${token}" in the ${side} part`, async () => {
+                        await driver.page.goto(`https://github.com/gorilla/mux/pull/117/files?diff=${diffType}`)
+                        // The browser extension takes a bit to initialize and register all event listeners.
+                        // Waiting here saves one retry cycle below in the common case.
+                        // If it's not enough, the retry will catch it.
+                        await driver.page.waitFor(1500)
+                        const tokenElement = await retry(async () => {
+                            const lineNumberElement = await driver.page.waitForSelector(`#${lineId}`, {
+                                timeout: 10000,
+                            })
+                            const row = (await driver.page.evaluateHandle(
+                                (element: Element) => element.closest('tr'),
+                                lineNumberElement
+                            ))!.asElement()!
```

Comment:
It would be nice if we could avoid these type assertions -- this could theoretically end up throwing with `Cannot call method asElement of undefined` instead of the more explicit `Expected row to exist` below.

With assertion functions, we could rewrite this as follows, and still get correct type narrowing after the `assert()` call:

```suggestion
                            ))?.asElement()
```

I think we're still blocked by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42786, though 😢

Reply:
Yeah I updated `@types/node` in this PR in the hopes that it would make use of the new assertion types, because it also annoyed me. I also checked `expect` and `chai` but they are not better. The cast from `asElement()` is definitely correct because we know the function can only return an element handle. The first cast is correct unless GitHub seriously changes its DOM so that IDs are not a descendant of a `tr` anymore (which I doubt will ever happen). Given that the error message is not _that_ much better I opted to save some code, but I'd definitely change it once https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42786 lands.